src/common/random_generator.h

```diff
@@ -0,0 +1,224 @@
+const int kCPUMinRndNumberPerThread = 64;
+// store how many global random states for CPU.
+const int kCPURndStateNum = 1024;
+
+template<typename DType>
+class RandGeneratorImpl<cpu, DType> {
+ public:
+  typedef typename std::conditional<std::is_floating_point<DType>::value,
+                                    DType, double>::type FType;
+
+  explicit RandGeneratorImpl<cpu, DType>(std::mt19937 *ptr_engine) : engine_(ptr_engine) {}
+
+  MSHADOW_XINLINE int rand() { return engine_->operator()(); }
+
+  MSHADOW_XINLINE FType uniform() {
+    typedef typename std::conditional<std::is_integral<DType>::value,
+                                      std::uniform_int_distribution<DType>,
+                                      std::uniform_real_distribution<FType>>::type GType;
+    GType dist_uniform;
+    return dist_uniform(*engine_);
+  }
+
+  MSHADOW_XINLINE FType normal() {
+    std::normal_distribution<FType> dist_normal;
+    return dist_normal(*engine_);
+  }
+
+ private:
+  std::mt19937 *engine_;
+};
+
+template<typename DType>
+class RandGenerator<cpu, DType> {
+ public:
+  MSHADOW_XINLINE void dispose() {}
+
+  MSHADOW_XINLINE RandGeneratorImpl<cpu, DType> Get(int idx = 0) {
+    std::mt19937 *ptr_engine = &states_[idx];
+    RandGeneratorImpl<cpu, DType> gen(ptr_engine);
+    return gen;
+  }
+
+  MSHADOW_XINLINE void Seed(Stream<cpu> *, uint32_t seed) {
+    for (int i = 0; i < kCPURndStateNum; ++i) (states_ + i)->seed(seed + i);
+  }
+
+  MSHADOW_XINLINE void set_state(int idx, const std::mt19937 &state) {
+    states_[idx] = state;
+  }
+
+ private:
+  std::mt19937 states_[kCPURndStateNum];
+};
+
+#if MXNET_USE_CUDA
+
+// at least how many random numbers should be generated by one GPU thread.
+const int kGPUMinRndNumberPerThread = 64;
+// store how many global random states for GPU.
+const int kGPURndStateNum = 32768;
+
+// uniform number generation in Cuda made consistent with stl (include 0 but exclude 1)
+// by using 1.0-curand_uniform().
+// Needed as some samplers in sampler.h won't be able to deal with
+// one of the boundary cases.
+template<typename DType>
+class RandGeneratorImpl<gpu, DType> {
+ public:
+  // Copy state to local memory for efficiency.
+  __device__ explicit RandGeneratorImpl(curandStatePhilox4_32_10_t *state)
+      : state_(*state) {}
+
+  MSHADOW_FORCE_INLINE __device__ int rand() {
+    return curand(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ float uniform() {
+    return static_cast<float>(1.0) - curand_uniform(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ float normal() {
+    return curand_normal(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ curandStatePhilox4_32_10_t get_state() {
+    return state_;
+  }
+
+ private:
+  curandStatePhilox4_32_10_t state_;
+};
+
+template<>
+class RandGeneratorImpl<gpu, double> {
+ public:
+  // Copy state to local memory for efficiency.
+  __device__ explicit RandGeneratorImpl(curandStatePhilox4_32_10_t *state)
+      : state_(*state) {}
```

Comment:
So you are copying state to state_ by value. Then wouldn't the next call of the same random operator give you the same results?

Reply:
I haven't tested the case for multiple runs of the generator. I should add that.

Reply:
Two choices:
- Follow the exact same patterns in CPU and GPU, i.e. copy by value but then also ensure to save the state again back in the RandGenerator<cpu> at the end of LaunchRNG (which isn't the case currently)
- Copy state by reference in the CPU case
I would prefer a consistent handling for both cases (i.e. first version).

Reply:
The implement here was right. It did save the state back in `LaunchRNG`.
I refactor it to a more readable version.